File: cpp/tp/pointers.cpp
#include <iostream>
using namespace std;


int main() {
  int  var1;
  char var2[10];


  // you can get the address of any variable with the & symbol
  // again, same as in C

  cout  << "Address of var1 variable: "
        << &var1 << endl;

  cout  << "Address of var1 variable: "
        << &var2 << endl;


  // pointers work as expected

  int year = 2021;
  int *pointer;
  int **pointer_pointer;

  pointer = &year; // address of integer year
  pointer_pointer = &pointer; // p -> p -> variable

  cout << "int year -> "       << year     << endl;
  cout << "int *pointer -> "   << pointer  << endl;
  cout << "deref *pointer -> " << *pointer << endl;

  // you can also get crazy in have pointer-pointers
  cout << "deref -> deref -> " << **pointer_pointer << endl;

  return 0;
}

File: fortran/tp/pointers.f90
program pointers
implicit none

  integer, pointer  :: my_pointer
  integer, target   :: first_target
  integer, target   :: second_target

  my_pointer => first_target
  my_pointer = 2021

  print *, "printing the a value from the same address"
  print *, "via the pointer ->  ", my_pointer
  print *, "via the variable -> ", first_target
  
  print *, "mutating the variable and printing again"

  first_target = first_target + 1
  print *, "via the pointer ->  ", my_pointer
  print *, "via the variable -> ", first_target
  
  print *, "mutating with a reassignment to value 1984"

  first_target = 1984
  print *, "via the pointer ->  ", my_pointer
  print *, "via the variable -> ", first_target
  
  print *, "reset pointer with `nullify`"

  nullify(my_pointer)

  print *, "the variable is still -> ", first_target

  print *, "set pointer to different variable"

  my_pointer => second_target

  print *, "query association with `associated` -> ", associated(my_pointer)
  print *, "querying is pointer is associated with a specific variable"
  print *, "pointer -> first variable?  ", associated(my_pointer, first_target)
  print *, "pointer -> second variable? ", associated(my_pointer, second_target)

  print *, "current values of pointer and second variable"
  print *, my_pointer
  print *, second_target

  print *, "changing the variable by assigning something to the pointer"
  
  my_pointer = 2001

  print *, my_pointer
  print *, second_target

end program pointers

